File: code/ch01/lambda_calculus.js
/* Church Encodings: Booleans */

const T = (a) => (b) => a; // True (in lambda calculus, λxy.x)
const F = (a) => (b) => b; // False (in lambda calculus, λxy.y)

const not = (p) => p(F)(T); // Negation
const and = (p) => (q) => p(q)(p); // Conjunction
const or = (p) => (q) => p(p)(q); // Disjunction
const eq = (p) => (q) => p(q)(not(q)); // Equal

const x = T;
const y = F;

console.log(not(x)); // !x
console.log(eq(x)(y)); // x == y
console.log(or(eq(not(x))(y))(and(y)(x))); // !x == y || (y && x)

module.exports = { T, F, not, and, or, eq };

File: code/ch01/lambda_calculus.py
# Church Encoding: Boolean

T = lambda a: lambda b: a # True (in lambda calculus, λxy.x)
T.__name__ = "True"
F = lambda a: lambda b: b # False (in lambda calculus, λxy.y)
F.__name__ = "False"

notB = lambda p: p(F)(T) # Negation
andB = lambda p: lambda q: p(q)(p) # Conjunction
orB = lambda p: lambda q: p(p)(q) # Disjunction
eqB = lambda p: lambda q: p(q)(notB(q)) # Equal

x = T
y = F

print(notB(x).__name__); # !x
print(eqB(x)(y).__name__); # x == y
print(orB(eqB(notB(x))(y))(andB(y)(x)).__name__); # !x == y || (y && x)
